### PR TITLE
COL-1215 Remove empty assignment categories if assignment goes away

### PR DESCRIPTION
**Pull request by paulkerschen**
https://jira.ets.berkeley.edu/jira/browse/COL-1215

### CODE REVIEW

File: node_modules/col-canvas/lib/poller.js [hunk 1]
```diff
@@ -553,9 +553,32 @@ var pollAssignments = function(ctx, users, opts, callback) {
...
+        // Re-fetch categories
+        CategoriesAPI.getCategories(ctx, true, true, function(err, categories) {
+          if (err) {
+            return callback(err);
+          }
+
+          // Remove any empty categories no longer corresponding to an active assignment. (The usual paranoid-mode delete isn't sufficient
+          // because that will prevent the category from being recreated if the assignment is later restored.)
+          var categoriesToDelete = _.filter(categories, function(c) {
+            return (c.canvas_assignment_id && (c.dataValues.asset_count === 0) && !(_.find(assignments, {'id': c.canvas_assignment_id})))
```
Comment (johncrossman) [inline node_modules/col-canvas/lib/poller.js]: @paulkerschen minor: unnecessary outer parens?
```
!(_.find(assignments, {'id': c.canvas_assignment_id}))
```

Comment (paulkerschen) [reply node_modules/col-canvas/lib/poller.js]: Reckon so, @johncrossman - amended.

File: node_modules/col-canvas/lib/poller.js [hunk 2]
```diff
@@ -553,9 +553,32 @@ var pollAssignments = function(ctx, users, opts, callback) {
...
+          });
+
+          async.eachSeries(categoriesToDelete, function(category, done) {
+            CategoriesAPI.deleteCategory(ctx, category.id, {'force': true}, function(err) {
+              if (err) {
+                log.error({'category': category.id, 'err': err}, 'Failed to delete a category');
+              }
+              return done();
+            });
+          }, callback);
```
Comment (johncrossman) [inline node_modules/col-canvas/lib/poller.js]: @paulkerschen  I'm very likely unclear...  In the previous impl, `callback` was invoked once: at finish of `eachSeries`.  In this new impl, we invoke once per iteration of outer `eachSeries`?

Comment (paulkerschen) [reply node_modules/col-canvas/lib/poller.js]: @johncrossman This is why async is awful. The outer `eachSeries` invokes `assignmentIterator` once per assignment, and the completion function (lines 558-579) afterward. The inner `eachSeries` invokes the iteration function (lines 573-578) once per category, and `callback` afterward.

**Comment by johncrossman** [comment]
I can't say I've missed Node...  Thanks @paulkerschen

**Pull request merged**